### PR TITLE
Facturas Cambios

### PR DESCRIPTION
**Pull request by Samara-MJ**
Cambios para la logica de Factura

### CODE REVIEW

**Review by acardenasl — APPROVED** [review]

File: ofipensiones/payments/models.py [hunk 1]
```diff
@@ -64,3 +66,23 @@ class Meta:
...
+class Factura(models.Model):
+    student = models.ForeignKey(Student, on_delete=models.DO_NOTHING)
+    total_amount = models.DecimalField(max_digits=10, decimal_places=2)
+    due_date = models.DateField()
+
+    def __str__(self) -> str:
+        return f"Factura for {self.student} - Total: {self.total_amount}"
+
+    
+class OfipensionesFactura(models.Model):
```
Comment (orion160) [inline ofipensiones/payments/models.py]: Remove, just use OfipensionesLog. It is a model for general logging.

File: ofipensiones/payments/models.py [hunk 2]
```diff
@@ -64,3 +66,23 @@ class Meta:
 
     def __str__(self) -> str:
         return f"{self.operation_name} - {self.timestamp}"
+    
+class Factura(models.Model):
```
Comment (orion160) [inline ofipensiones/payments/models.py]: Translate into invoice or bill

File: ofipensiones/payments/serializers.py [hunk 3]
```diff
@@ -28,3 +28,22 @@ def get_school_info(self, obj):
             "city": obj.student.school.city,
             "address": obj.student.school.address,
         }
+    
+
+class FacturaSerializer(serializers.ModelSerializer):
```
Comment (orion160) [inline ofipensiones/payments/serializers.py]: Translate

File: ofipensiones/payments/urls.py [hunk 4]
```diff
@@ -1,6 +1,9 @@
 from django.urls import path
-from .views import PaymentReceiptList
+from .views import (PaymentReceiptList,
+                    FacturaList
+)
 
 urlpatterns = [
     path("payment_receipt/", PaymentReceiptList.as_view(), name="payment_receipt_list"),
+    path("factura/", FacturaList.as_view(), name="factura_list"),
```
Comment (orion160) [inline ofipensiones/payments/urls.py]: Translate

File: ofipensiones/payments/views.py [hunk 5]
```diff
@@ -20,3 +25,90 @@ def get(self, request):
         OfipensionesLog(operation_name="payment_receipts", time_taken=delta).save()
 
         return Response(serializer.data, status=status.HTTP_200_OK)
+    
+
+
+class FacturaList(APIView):
```
Comment (orion160) [inline ofipensiones/payments/views.py]: Apply changes

Comment (orion160) [inline ofipensiones/payments/views.py]: Apply CRUD also on PaymentReceipt

**Review by orion160 — CHANGES_REQUESTED** [review]

File: ofipensiones/payments/views.py [hunk 6]
```diff
@@ -20,3 +25,90 @@ def get(self, request):
...
+            e = time.time()
+            
+            delta = timedelta(seconds=e - s)
+            OfipensionesFactura(operation_name="create_factura", time_taken=delta).save()
+
+            return Response(serializer.data, status=status.HTTP_201_CREATED)
+
+        e = time.time()
+        delta = timedelta(seconds=e - s)
+        OfipensionesFactura(operation_name="create_factura_failed", time_taken=delta).save()
```
Comment (orion160) [inline ofipensiones/payments/views.py]: Translate

Comment (orion160) [reply ofipensiones/payments/views.py]: Also good use of logs

File: ofipensiones/payments/views.py [hunk 7]
```diff
@@ -20,3 +25,90 @@ def get(self, request):
         OfipensionesLog(operation_name="payment_receipts", time_taken=delta).save()
```
Comment (orion160) [inline ofipensiones/payments/views.py]: append payment_receipts_<operation_name>

as example:

payment_receipts_get

File: ofipensiones/payments/views.py [hunk 8]
```diff
@@ -20,3 +25,90 @@ def get(self, request):
...
+            raise Http404
+
+    def get(self, request, pk, format=None):
+        s = time.time()
+        factura = self.get_object(pk)
+        serializer = FacturaSerializer(factura)
+        e = time.time()
+
+        delta = timedelta(seconds=e - s)
+        OfipensionesFactura(operation_name="retrieve_factura", time_taken=delta).save()
```
Comment (orion160) [inline ofipensiones/payments/views.py]: Translate

File: ofipensiones/payments/views.py [hunk 9]
```diff
@@ -20,3 +25,90 @@ def get(self, request):
...
+            e = time.time()
+
+            delta = timedelta(seconds=e - s)
+            OfipensionesFactura(operation_name="update_factura", time_taken=delta).save()
+
+            return Response(serializer.data, status=status.HTTP_200_OK)
+
+        e = time.time()
+        delta = timedelta(seconds=e - s)
+        OfipensionesFactura(operation_name="update_factura_failed", time_taken=delta).save()
```
Comment (orion160) [inline ofipensiones/payments/views.py]: bill_update
bill_create
bill_update_failed
...

**Review by orion160 — CHANGES_REQUESTED** [review]

**Review by orion160 — APPROVED** [review]

**Pull request merged**